File: flut/.flutter/lib/flutter/material/theme_data.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/error.dart';

class FlutMaterialTapTargetSize extends FlutEnumObject<MaterialTapTargetSize> {
  const FlutMaterialTapTargetSize()
    : super('MaterialTapTargetSize', const {
        'padded': MaterialTapTargetSize.padded,
        'shrinkWrap': MaterialTapTargetSize.shrinkWrap,
      });
}

class FlutVisualDensity extends FlutValueObject {
  final VisualDensity value;
  const FlutVisualDensity(this.value) : super('VisualDensity');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['horizontal'] = FlutDouble.encode(value.horizontal);
    result['vertical'] = FlutDouble.encode(value.vertical);
    return result;
  }

  static VisualDensity? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return VisualDensity(
      horizontal: runtime.unpackRequiredField<double>(data, 'horizontal'),
      vertical: runtime.unpackRequiredField<double>(data, 'vertical'),
    );
  }
}

class FlutThemeData with FlutRealtimeObject<ThemeData> {
  FlutThemeData.createFromData({
    required FlutRuntime runtime,
    required Map<String, dynamic> data,
    required ThemeData target,
  }) {
    initRealtimeFromData(runtime: runtime, data: data, target: target);
  }

  FlutThemeData.createFromObject({
    required FlutRuntime runtime,
    required int oid,
    required ThemeData target,
  }) {
    initRealtimeFromObject(
      runtime: runtime,
      oid: oid,
      type: 'ThemeData',
      target: target,
    );
  }

  static FlutThemeData flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FlutThemeData.createFromData(
      runtime: runtime,
      data: data,
      target: ThemeData(
        colorScheme: runtime.unpackOptionalField<ColorScheme>(
          data,
          'colorScheme',
        ),
        brightness: runtime.unpackOptionalField<Brightness>(data, 'brightness'),
        colorSchemeSeed: runtime.unpackOptionalField<Color>(
          data,
          'colorSchemeSeed',
        ),
        useMaterial3: runtime.unpackOptionalField<bool>(data, 'useMaterial3'),
        textTheme: runtime.unpackOptionalField<TextTheme>(data, 'textTheme'),
        primaryTextTheme: runtime.unpackOptionalField<TextTheme>(
          data,
          'primaryTextTheme',
        ),
        fontFamily: runtime.unpackOptionalField<String>(data, 'fontFamily'),
        fontFamilyFallback: runtime.unpackOptionalField<List<String>>(
          data,
          'fontFamilyFallback',
        ),
        primaryColor: runtime.unpackOptionalField<Color>(data, 'primaryColor'),
        primaryColorDark: runtime.unpackOptionalField<Color>(
          data,
          'primaryColorDark',
        ),
        primaryColorLight: runtime.unpackOptionalField<Color>(
          data,
          'primaryColorLight',
        ),
        canvasColor: runtime.unpackOptionalField<Color>(data, 'canvasColor'),
        cardColor: runtime.unpackOptionalField<Color>(data, 'cardColor'),
        scaffoldBackgroundColor: runtime.unpackOptionalField<Color>(
          data,
          'scaffoldBackgroundColor',
        ),
        dividerColor: runtime.unpackOptionalField<Color>(data, 'dividerColor'),
        focusColor: runtime.unpackOptionalField<Color>(data, 'focusColor'),
        highlightColor: runtime.unpackOptionalField<Color>(
          data,
          'highlightColor',
        ),
        hintColor: runtime.unpackOptionalField<Color>(data, 'hintColor'),
        hoverColor: runtime.unpackOptionalField<Color>(data, 'hoverColor'),
        shadowColor: runtime.unpackOptionalField<Color>(data, 'shadowColor'),
        splashColor: runtime.unpackOptionalField<Color>(data, 'splashColor'),
        disabledColor: runtime.unpackOptionalField<Color>(
          data,
          'disabledColor',
        ),
        unselectedWidgetColor: runtime.unpackOptionalField<Color>(
          data,
          'unselectedWidgetColor',
        ),
        secondaryHeaderColor: runtime.unpackOptionalField<Color>(
          data,
          'secondaryHeaderColor',
        ),
        materialTapTargetSize: runtime
            .unpackOptionalField<MaterialTapTargetSize>(
              data,
              'materialTapTargetSize',
            ),
        platform: runtime.unpackOptionalField<TargetPlatform>(data, 'platform'),
        applyElevationOverlayColor: runtime.unpackOptionalField<bool>(
          data,
          'applyElevationOverlayColor',
        ),
        useSystemColors: runtime.unpackOptionalField<bool>(
          data,
          'useSystemColors',
        ),
      ),
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'brightness':
        return flutTarget.brightness;
      case 'colorScheme':
        return flutTarget.colorScheme;
      case 'useMaterial3':
        return flutTarget.useMaterial3;
      case 'textTheme':
        return flutTarget.textTheme;
      case 'primaryColor':
        return flutTarget.primaryColor;
      case 'primaryColorDark':
        return flutTarget.primaryColorDark;
      case 'primaryColorLight':
        return flutTarget.primaryColorLight;
      case 'canvasColor':
        return flutTarget.canvasColor;
      case 'cardColor':
        return flutTarget.cardColor;
      case 'scaffoldBackgroundColor':
        return flutTarget.scaffoldBackgroundColor;
      case 'dividerColor':
        return flutTarget.dividerColor;
      case 'focusColor':
        return flutTarget.focusColor;
      case 'highlightColor':
        return flutTarget.highlightColor;
      case 'hintColor':
        return flutTarget.hintColor;
      case 'hoverColor':
        return flutTarget.hoverColor;
      case 'shadowColor':
        return flutTarget.shadowColor;
      case 'splashColor':
        return flutTarget.splashColor;
      case 'disabledColor':
        return flutTarget.disabledColor;
      case 'unselectedWidgetColor':
        return flutTarget.unselectedWidgetColor;
      case 'secondaryHeaderColor':
        return flutTarget.secondaryHeaderColor;
      case 'primaryTextTheme':
        return flutTarget.primaryTextTheme;
      case 'iconTheme':
        return flutTarget.iconTheme;
      case 'primaryIconTheme':
        return flutTarget.primaryIconTheme;
      case 'platform':
        return flutTarget.platform;
      case 'materialTapTargetSize':
        return flutTarget.materialTapTargetSize;
      case 'applyElevationOverlayColor':
        return flutTarget.applyElevationOverlayColor;
      case 'visualDensity':
        return flutTarget.visualDensity;
      case 'typography':
        return flutTarget.typography;
    }
    throw FlutUnknownPropertyException('ThemeData', property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    throw FlutUnknownPropertyException('ThemeData', property);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    throw FlutUnknownMethodException(method);
  }
}

File: flut/flutter/material/theme_data.py
from typing import Optional, override
from flut._flut.engine import (
    FlutEnumObject,
    FlutRealtimeObject,
    FlutValueObject,
    _flut_pack_value,
)
from flut._flut.runtime import _flut_unpack_optional_field
from flut.dart.ui import Brightness, Color, _Offset
from flut.flutter.material.color_scheme import ColorScheme
from flut.flutter.material.ink_well import InteractiveInkFeatureFactory
from flut.flutter.material.text_theme import TextTheme


class MaterialTapTargetSize(FlutEnumObject):
    padded: "MaterialTapTargetSize"
    shrinkWrap: "MaterialTapTargetSize"


class _VisualDensity(FlutValueObject):
    _flut_type = "VisualDensity"

    def __init__(self, *, horizontal: float = 0.0, vertical: float = 0.0):
        super().__init__()
        self.horizontal = horizontal
        self.vertical = vertical

    @property
    def baseSizeAdjustment(self):
        return _Offset(self.horizontal * 4.0, self.vertical * 4.0)

    @staticmethod
    def _flut_unpack(data: dict) -> "_VisualDensity":
        return _VisualDensity(
            horizontal=_flut_unpack_optional_field(data, "horizontal") or 0.0,
            vertical=_flut_unpack_optional_field(data, "vertical") or 0.0,
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["horizontal"] = _flut_pack_value(self.horizontal)
        result["vertical"] = _flut_pack_value(self.vertical)
        return result


class VisualDensity(_VisualDensity):
    minimumDensity = -4.0
    maximumDensity = 4.0
    comfortable = _VisualDensity(horizontal=-1.0, vertical=-1.0)
    standard = _VisualDensity(horizontal=0.0, vertical=0.0)
    compact = _VisualDensity(horizontal=-2.0, vertical=-2.0)


class ThemeData(FlutRealtimeObject):
    _flut_type = "ThemeData"

    def __init__(
        self,
        *,
        colorScheme=None,
        brightness: Optional[Brightness] = None,
        colorSchemeSeed: Optional[Color] = None,
        useMaterial3: Optional[bool] = None,
        textTheme=None,
        fontFamily: Optional[str] = None,
        fontFamilyFallback: Optional[list[str]] = None,
        package: Optional[str] = None,
        primaryColor: Optional[Color] = None,
        primaryColorDark: Optional[Color] = None,
        primaryColorLight: Optional[Color] = None,
        canvasColor: Optional[Color] = None,
        cardColor: Optional[Color] = None,
        scaffoldBackgroundColor: Optional[Color] = None,
        dividerColor: Optional[Color] = None,
        focusColor: Optional[Color] = None,
        highlightColor: Optional[Color] = None,
        hintColor: Optional[Color] = None,
        hoverColor: Optional[Color] = None,
        shadowColor: Optional[Color] = None,
        splashColor: Optional[Color] = None,
        disabledColor: Optional[Color] = None,
        unselectedWidgetColor: Optional[Color] = None,
        secondaryHeaderColor: Optional[Color] = None,
        iconTheme=None,
        primaryIconTheme=None,
        primaryTextTheme=None,
        typography=None,
        applyElevationOverlayColor: Optional[bool] = None,
        materialTapTargetSize=None,
        pageTransitionsTheme=None,
        platform=None,
        splashFactory: Optional[InteractiveInkFeatureFactory] = None,
        useSystemColors: Optional[bool] = None,
        visualDensity=None,
        inputDecorationTheme=None,
    ):
        super().__init__()
        props = {}
        if colorScheme is not None:
            props["colorScheme"] = colorScheme._flut_pack()
        if brightness is not None:
            props["brightness"] = brightness._flut_pack()
        if colorSchemeSeed is not None:
            props["colorSchemeSeed"] = colorSchemeSeed._flut_pack()
        if useMaterial3 is not None:
            props["useMaterial3"] = useMaterial3
        if textTheme is not None:
            props["textTheme"] = textTheme._flut_pack()
        if fontFamily is not None:
            props["fontFamily"] = fontFamily
        if fontFamilyFallback is not None:
            props["fontFamilyFallback"] = fontFamilyFallback
        if package is not None:
            props["package"] = package
        if primaryColor is not None:
            props["primaryColor"] = primaryColor._flut_pack()
        if primaryColorDark is not None:
            props["primaryColorDark"] = primaryColorDark._flut_pack()
        if primaryColorLight is not None:
            props["primaryColorLight"] = primaryColorLight._flut_pack()
        if canvasColor is not None:
            props["canvasColor"] = canvasColor._flut_pack()
        if cardColor is not None:
            props["cardColor"] = cardColor._flut_pack()
        if scaffoldBackgroundColor is not None:
            props["scaffoldBackgroundColor"] = scaffoldBackgroundColor._flut_pack()
        if dividerColor is not None:
            props["dividerColor"] = dividerColor._flut_pack()
        if focusColor is not None:
            props["focusColor"] = focusColor._flut_pack()
        if highlightColor is not None:
            props["highlightColor"] = highlightColor._flut_pack()
        if hintColor is not None:
            props["hintColor"] = hintColor._flut_pack()
        if hoverColor is not None:
            props["hoverColor"] = hoverColor._flut_pack()
        if shadowColor is not None:
            props["shadowColor"] = shadowColor._flut_pack()
        if splashColor is not None:
            props["splashColor"] = splashColor._flut_pack()
        if disabledColor is not None:
            props["disabledColor"] = disabledColor._flut_pack()
        if unselectedWidgetColor is not None:
            props["unselectedWidgetColor"] = unselectedWidgetColor._flut_pack()
        if secondaryHeaderColor is not None:
            props["secondaryHeaderColor"] = secondaryHeaderColor._flut_pack()
        if iconTheme is not None:
            props["iconTheme"] = _flut_pack_value(iconTheme)
        if primaryIconTheme is not None:
            props["primaryIconTheme"] = _flut_pack_value(primaryIconTheme)
        if primaryTextTheme is not None:
            props["primaryTextTheme"] = primaryTextTheme._flut_pack()
        if typography is not None:
            props["typography"] = _flut_pack_value(typography)
        if applyElevationOverlayColor is not None:
            props["applyElevationOverlayColor"] = applyElevationOverlayColor
        if materialTapTargetSize is not None:
            props["materialTapTargetSize"] = _flut_pack_value(materialTapTargetSize)
        if pageTransitionsTheme is not None:
            props["pageTransitionsTheme"] = _flut_pack_value(pageTransitionsTheme)
        if platform is not None:
            props["platform"] = _flut_pack_value(platform)
        if splashFactory is not None:
            props["splashFactory"] = _flut_pack_value(splashFactory)
        if useSystemColors is not None:
            props["useSystemColors"] = useSystemColors
        if visualDensity is not None:
            props["visualDensity"] = _flut_pack_value(visualDensity)
        if inputDecorationTheme is not None:
            props["inputDecorationTheme"] = _flut_pack_value(inputDecorationTheme)
        self._flut_create(props=props)

    @property
    def brightness(self) -> Brightness:
        return self._flut_get("brightness")

    @property
    def colorScheme(self) -> ColorScheme:
        return self._flut_get("colorScheme")

    @property
    def useMaterial3(self) -> bool:
        return self._flut_get("useMaterial3")

    @property
    def textTheme(self) -> TextTheme:
        return self._flut_get("textTheme")

    @property
    def primaryColor(self) -> Color:
        return self._flut_get("primaryColor")

    @property
    def primaryColorDark(self) -> Color:
        return self._flut_get("primaryColorDark")

    @property
    def primaryColorLight(self) -> Color:
        return self._flut_get("primaryColorLight")

    @property
    def canvasColor(self) -> Color:
        return self._flut_get("canvasColor")

    @property
    def cardColor(self) -> Color:
        return self._flut_get("cardColor")

    @property
    def scaffoldBackgroundColor(self) -> Color:
        return self._flut_get("scaffoldBackgroundColor")

    @property
    def dividerColor(self) -> Color:
        return self._flut_get("dividerColor")

    @property
    def focusColor(self) -> Color:
        return self._flut_get("focusColor")

    @property
    def highlightColor(self) -> Color:
        return self._flut_get("highlightColor")

    @property
    def hintColor(self) -> Color:
        return self._flut_get("hintColor")

    @property
    def hoverColor(self) -> Color:
        return self._flut_get("hoverColor")

    @property
    def shadowColor(self) -> Color:
        return self._flut_get("shadowColor")

    @property
    def splashColor(self) -> Color:
        return self._flut_get("splashColor")

    @property
    def disabledColor(self) -> Color:
        return self._flut_get("disabledColor")

    @property
    def unselectedWidgetColor(self) -> Color:
        return self._flut_get("unselectedWidgetColor")

    @property
    def secondaryHeaderColor(self) -> Color:
        return self._flut_get("secondaryHeaderColor")

    @property
    def primaryTextTheme(self) -> TextTheme:
        return self._flut_get("primaryTextTheme")
